langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/WeaviateEmbeddingStoreImpl.java

```diff
@@ -0,0 +1,192 @@
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import io.weaviate.client.Config;
+import io.weaviate.client.WeaviateAuthClient;
+import io.weaviate.client.WeaviateClient;
+import io.weaviate.client.v1.data.model.WeaviateObject;
+import io.weaviate.client.v1.data.replication.model.ConsistencyLevel;
+import io.weaviate.client.v1.graphql.query.argument.NearVectorArgument;
+import io.weaviate.client.v1.graphql.query.fields.Field;
+import java.security.MessageDigest;
+import java.util.*;
+import java.util.stream.Collectors;
+import lombok.Builder;
+import lombok.SneakyThrows;
+import lombok.val;
+
+public class WeaviateEmbeddingStoreImpl implements EmbeddingStore<TextSegment> {
+
+  private static final String DEFAULT_CLASS = "Default";
+  private static final Double DEFAULT_MIN_CERTAINTY = 0.8;
+  private static final String METADATA_TEXT_SEGMENT = "text";
+  private static final String ADDITIONALS = "_additional";
+
+  private final WeaviateClient client;
+  private final String objectClass;
+
+  @Builder
+  @SneakyThrows
+  public WeaviateEmbeddingStoreImpl(String apiKey, String scheme, String host, String objectClass) {
+    client = WeaviateAuthClient.apiKey(new Config(scheme, host), apiKey);
+    this.objectClass = objectClass != null ? objectClass : DEFAULT_CLASS;
+  }
+
+  @Override
+  public String add(Embedding embedding) {
+    val id = generateRandomId();
+    add(id, embedding);
+    return id;
+  }
+
+  @Override
+  public void add(String id, Embedding embedding) {
+    addAll(singletonList(id), singletonList(embedding), null);
+  }
+
+  @Override
+  public String add(Embedding embedding, TextSegment textSegment) {
+    return addAll(singletonList(embedding), textSegment == null ? null : singletonList(textSegment))
+      .stream()
+      .findFirst()
+      .orElse(null);
+  }
+
+  @Override
+  public List<String> addAll(List<Embedding> embeddings) {
+    return addAll(embeddings, null);
+  }
+
+  @Override
+  public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+    return addAll(null, embeddings, embedded);
+  }
+
+  @Override
+  public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults) {
+    return findRelevant(referenceEmbedding, maxResults, DEFAULT_MIN_CERTAINTY);
+  }
+
+  @Override
+  public List<EmbeddingMatch<TextSegment>> findRelevant(
+    Embedding referenceEmbedding,
+    int maxResults,
+    double minCertainty
+  ) {
+    val result = client
+      .graphQL()
+      .get()
+      .withClassName(objectClass)
+      .withFields(
+        Field.builder().name(METADATA_TEXT_SEGMENT).build(),
+        Field
+          .builder()
+          .name(ADDITIONALS)
+          .fields(
+            Field.builder().name("id").build(),
+            Field.builder().name("certainty").build(),
```

Comment:
Why do we need to store and retrieve certainty? As far as I understand, this is a dynamic parameter that is defined during search time.

Reply:
no, all of them under additional (id, certainty and vector) are Weaviate internals, they are mentioned here to be shown in search results